tiledb/sm/filesystem/s3/AWSCredentialsProviderChain.cc

```diff
@@ -0,0 +1,166 @@
+  AWS_LOGSTREAM_DEBUG(
+      DefaultCredentialsProviderChainTag,
+      "The environment variable value "
+          << GeneralHTTPCredentialsProvider::
+                 AWS_CONTAINER_CREDENTIALS_RELATIVE_URI
+          << " is " << relativeUri);
+
+  const auto absoluteUri = Aws::Environment::GetEnv(
+      GeneralHTTPCredentialsProvider::AWS_CONTAINER_CREDENTIALS_FULL_URI);
+  AWS_LOGSTREAM_DEBUG(
+      DefaultCredentialsProviderChainTag,
+      "The environment variable value "
+          << GeneralHTTPCredentialsProvider::AWS_CONTAINER_CREDENTIALS_FULL_URI
+          << " is " << absoluteUri);
+
+  const auto ec2MetadataDisabled =
+      Aws::Environment::GetEnv(AWS_EC2_METADATA_DISABLED);
+  AWS_LOGSTREAM_DEBUG(
+      DefaultCredentialsProviderChainTag,
+      "The environment variable value " << AWS_EC2_METADATA_DISABLED << " is "
+                                        << ec2MetadataDisabled);
+
+  if (!relativeUri.empty() || !absoluteUri.empty()) {
+    const Aws::String token = Aws::Environment::GetEnv(
+        GeneralHTTPCredentialsProvider::AWS_CONTAINER_AUTHORIZATION_TOKEN);
+    const Aws::String tokenPath = Aws::Environment::GetEnv(
+        GeneralHTTPCredentialsProvider::AWS_CONTAINER_AUTHORIZATION_TOKEN_FILE);
+
+    auto genProvider =
+        clientConfig ? make_shared<GeneralHTTPCredentialsProvider>(
+                           HERE(),
+                           *clientConfig,
+                           relativeUri,
+                           absoluteUri,
+                           token,
+                           tokenPath) :
+                       make_shared<GeneralHTTPCredentialsProvider>(
+                           HERE(), relativeUri, absoluteUri, token, tokenPath);
+    if (genProvider && genProvider->IsValid()) {
+      AddProvider(std::move(genProvider));
+      auto& uri = !relativeUri.empty() ? relativeUri : absoluteUri;
+      AWS_LOGSTREAM_INFO(
+          DefaultCredentialsProviderChainTag,
+          "Added General HTTP / ECS credentials provider with ur: ["
+              << uri << "] to the provider chain with a"
+              << ((token.empty() && tokenPath.empty()) ? "n empty " :
+                                                         " non-empty ")
+              << "authorization token.");
+    } else {
+      AWS_LOGSTREAM_ERROR(
+          DefaultCredentialsProviderChainTag,
+          "Unable to create GeneralHTTPCredentialsProvider");
+    }
+  } else if (
+      Aws::Utils::StringUtils::ToLower(ec2MetadataDisabled.c_str()) != "true") {
+    auto ec2MetadataClient = clientConfig ?
+                                 make_shared<Aws::Internal::EC2MetadataClient>(
+                                     HERE(), *clientConfig) :
+                                 nullptr;
+    AddProvider(make_shared<InstanceProfileCredentialsProvider>(
+        HERE(),
+        make_shared<Aws::Config::EC2InstanceProfileConfigLoader>(
+            std::move(ec2MetadataClient))));
```

Comment:
On second thought it might not be worth it to pass `ClientConfiguration` here; setting non-default options might change something I cannot think of and the IMDS connection happens with HTTP for which we don't need the CA path either way.

Reply:
I can see the arguments both ways, I'd lean in favor of supporting passing it for consistency.